src/sage/calculus/calculus.py

```diff
@@ -1410,76 +1493,159 @@ def limit(ex, dir=None, taylor=False, algorithm='maxima', **argv):
+    elif effective_algorithm == 'maxima_taylor':
         if dir is None:
             l = maxima.sr_tlimit(ex, v, a)
         elif dir in dir_plus:
             l = maxima.sr_tlimit(ex, v, a, 'plus')
         elif dir in dir_minus:
             l = maxima.sr_tlimit(ex, v, a, 'minus')
-    elif algorithm == 'sympy':
+    elif effective_algorithm == 'sympy':
         import sympy
-        if dir is None:
-            l = sympy.limit(ex._sympy_(), v._sympy_(), a._sympy_())
-        elif dir in dir_plus:
-            l = sympy.limit(ex._sympy_(), v._sympy_(), a._sympy_(), dir='+')
+        sympy_dir = '+-'  
+        if dir in dir_plus:
+            sympy_dir = '+'
         elif dir in dir_minus:
-            l = sympy.limit(ex._sympy_(), v._sympy_(), a._sympy_(), dir='-')
-    elif algorithm == 'fricas':
+            sympy_dir = '-'
+        l = sympy.limit(ex._sympy_(), v._sympy_(), a._sympy_(), dir=sympy_dir)
+    elif effective_algorithm == 'fricas':
         from sage.interfaces.fricas import fricas
         eq = fricas.equation(v._fricas_(), a._fricas_())
         f = ex._fricas_()
-        if dir is None:
-            l = fricas.limit(f, eq).sage()
-            if isinstance(l, dict):
-                l = maxima("und")
-        elif dir in dir_plus:
-            l = fricas.limit(f, eq, '"right"').sage()
+        fricas_dir_arg = None
+        if dir in dir_plus:
+            fricas_dir_arg = '"right"'
         elif dir in dir_minus:
-            l = fricas.limit(f, eq, '"left"').sage()
-    elif algorithm == 'giac':
+            fricas_dir_arg = '"left"'
+
+        if fricas_dir_arg:
+             l = fricas.limit(f, eq, fricas_dir_arg).sage()
+        else:
+             l_raw = fricas.limit(f, eq).sage()
+             if isinstance(l_raw, dict):
+                 l = SR('und')
+             else:
+                 l = l_raw
+    elif effective_algorithm == 'giac':
         from sage.libs.giac.giac import libgiac
-        v = v._giac_init_()
-        a = a._giac_init_()
-        if dir is None:
-            l = libgiac.limit(ex, v, a).sage()
-        elif dir in dir_plus:
-            l = libgiac.limit(ex, v, a, 1).sage()
+        giac_v = v._giac_init_()
+        giac_a = a._giac_init_()
+        giac_dir_arg = 0 # Default for two-sided 
+        if dir in dir_plus:
+            giac_dir_arg = 1
         elif dir in dir_minus:
-            l = libgiac.limit(ex, v, a, -1).sage()
-    elif algorithm == 'mathematica_free':
-        return mma_free_limit(ex, v, a, dir)
+            giac_dir_arg = -1
+        l = libgiac.limit(ex, giac_v, giac_a, giac_dir_arg).sage()
+    elif effective_algorithm == 'mathematica_free':
+        # Ensuring mma_free_limit exists 
+        l = mma_free_limit(ex, v, a, dir)
     else:
-        raise ValueError("Unknown algorithm: %s" % algorithm)
-    return ex.parent()(l)
+        raise ValueError("Unknown algorithm: %s" % effective_algorithm)
 
+    original_parent = ex.parent()
+
+    return original_parent(l)
```

Comment:
```suggestion
    return ex.parent()(l)
```

I think the original one line here is a bit nicer since we aren't using `original_parent` elsewhere. `ex.parent()(l)` might be negligibly faster too, although I'm not familiar enough with the Python interpreter to say that 100% for sure.